holonote/app/panel.py

```diff
@@ -83,41 +84,94 @@ def _create_visible_widget(self):
-            colormap = dict(zip(data, _default_color))
-        else:
-            colormap = style._colormap
-        if isinstance(colormap, dict):
-            stylesheet = """
-            option:after {
-              content: "";
-              width: 10px;
-              height: 10px;
-              position: absolute;
-              border-radius: 50%;
-              left: calc(100% - var(--design-unit, 4) * 2px - 3px);
-              border: 1px solid black;
-              opacity: 0.5;
-            }"""
-            for i, color in enumerate(colormap.values()):
-                stylesheet += f"""
-            option:nth-child({i + 1}):after {{
-                background-color: {color};
-            }}"""
+            self.colormap = dict(zip(data, _default_color))
         else:
-            stylesheet = ""
+            self.colormap = style._colormap or {}
 
-        options = list(colormap)
+        all_options = set(self.annotator.df[self.annotator.groupby].unique())
+        for option in all_options:
+            if option not in self.colormap:
+                self.colormap[option] = (
+                    style.color or _default_color[len(self.colormap) % len(_default_color)]
+                )
+        self._update_stylesheet()
+
+        options = sorted(self.colormap.keys())
         self.visible_widget = pn.widgets.MultiSelect(
             name="Visible",
             options=options,
             value=self.annotator.visible or options,
-            stylesheets=[stylesheet],
+            stylesheets=[self.stylesheet],
         )
         self.annotator.visible = self.visible_widget
 
+    def _update_stylesheet(self):
+        self.stylesheet = """
+        option:after {
+          content: "";
+          width: 8px;
+          height: 8px;
+          position: absolute;
+          border-radius: 50%;
+          left: 1px;
+          border: 1px solid black;
+          opacity: 0.5;
+        }"""
+        for _, (option, color) in enumerate(sorted(self.colormap.items())):
+            self.stylesheet += f"""
+        option[value="{option}"]:after {{
+            background-color: {color};
+        }}"""
+
+    def _get_unique_color(self):
+        used_colors = set(self.colormap.values())
+        for color in _default_color:
+            if color not in used_colors:
+                return color
+        # If all default colors are used, cycle through again
+        return _default_color[len(self.colormap) % len(_default_color)]
+
+    def _update_visible_widget(self, event):
+        if event.type == "create":
+            new_option = event.fields[self.annotator.groupby]
+            old_options = list(self.visible_widget.options)
+            old_values = list(self.visible_widget.value)
+            if new_option not in old_options:
+                self.visible_widget.options = sorted([*old_options, new_option])
+                self.visible_widget.value = [*old_values, new_option]
```

Comment:
Always better to update all parameters at once.

```suggestion
                self.visible_widget.param.update(
                    options=sorted([*old_options, new_option]),
                    value=[*old_values, new_option]
                )
```

The same applies below.

Reply:
thanks for the tip, I committed this change in this particular 'create' event type. The other place where two parameters are updated is under the 'update' event, but I will keep that separate because if a group (e.g. 'A') had been set to not be visible, and then an annotation that is the last of it's kind is updated (e.g. from 'B' to 'A'), then I would want the group ('A') to remain not visible, even though we now have to update the 'options' parameter. That is why I am keeping them separate for that particular event type.